pkg/api/server/v1alpha2/auth/impersonation/impersonation.go

```diff
@@ -0,0 +1,217 @@
+			i.resourceAttributes = append(i.resourceAttributes, authorizationv1.ResourceAttributes{
+				Group:       authenticationv1.SchemeGroupVersion.Group,
+				Name:        value,
+				Resource:    "userextras",
+				Subresource: extraKey,
+				Verb:        "impersonate",
+			})
+		}
+	}
+
+	// When impersonating a non-anonymous user, include the 'system:authenticated' group in the impersonated user info:
+	// - if no groups were specified
+	// - if a group has been specified other than 'system:authenticated'
+	// If 'system:unauthenticated' group has been specified we should not include the 'system:authenticated' group.
+	if hasUser {
+		if i.userInfo.Name != user.Anonymous {
+			addAuthenticated := true
+			for _, group := range i.userInfo.Groups {
+				if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+					addAuthenticated = false
+					break
+				}
+			}
+
+			if addAuthenticated {
+				i.userInfo.Groups = append(i.userInfo.Groups, user.AllAuthenticated)
+			}
+		} else {
+			addUnauthenticated := true
+			for _, group := range i.userInfo.Groups {
+				if group == user.AllUnauthenticated {
+					addUnauthenticated = false
+					break
+				}
+			}
+
+			if addUnauthenticated {
+				i.userInfo.Groups = append(i.userInfo.Groups, user.AllUnauthenticated)
+			}
+		}
+	} else {
+		i.userInfo = nil
+
+		// Impersonate-User header is mandatory.
+		// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation
+		if hasGroups || hasExtra || hasUID {
+			return ErrorImpersonateUserRequired
+		} else {
+			return ErrorNoImpersonationData
+		}
+	}
+
+	return nil
+}
+
+func unescapeExtraKey(encodedKey string) string {
+	key, err := url.PathUnescape(encodedKey) // Decode %-encoded bytes.
+	if err != nil {
+		return encodedKey // Always record extra strings, even if malformed/unencoded.
+	}
+	return key
+}
+
+// Check checks if the requester has access to impersonate every resource.
+func (i *Impersonation) Check(ctx context.Context, authorizer authorizationclient.AuthorizationV1Interface, requester string) error {
+	if i.resourceAttributes == nil {
+		return ErrorNoImpersonationData
+	}
+	for _, resourceAttribute := range i.resourceAttributes {
+		sar, err := authorizer.SubjectAccessReviews().Create(ctx, &authorizationv1.SubjectAccessReview{
+			Spec: authorizationv1.SubjectAccessReviewSpec{
+				User: requester,
+				ResourceAttributes: &authorizationv1.ResourceAttributes{
+					Group:       resourceAttribute.Group,
+					Name:        resourceAttribute.Name,
+					Namespace:   resourceAttribute.Namespace,
+					Resource:    resourceAttribute.Resource,
+					Subresource: resourceAttribute.Subresource,
+					Verb:        resourceAttribute.Verb,
+				},
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		if !sar.Status.Allowed {
+			return fmt.Errorf("forbidden: '%s' doesn't have access to impersonate '%s'", requester, resourceAttribute.Resource)
```

Comment:
small change to error message:

```suggestion
			return fmt.Errorf("forbidden: '%s' doesn't have permission to impersonate '%s'", requester, resourceAttribute.Resource)
```

Reply:
:thumbsup: Changed